tracer/src/Datadog.Trace/SpanContext.cs

```diff
@@ -438,37 +438,10 @@ internal string PrepareTagsHeaderForPropagation()
         /// <param name="edgeTags">The edge tags for this checkpoint. NOTE: These MUST be sorted alphabetically</param>
         /// <param name="payloadSizeBytes">Payload size in bytes</param>
         /// <param name="timeInQueueMs">Edge start time extracted from the message metadata. Used only if this is start of the pathway</param>
-        internal void SetCheckpoint(DataStreamsManager manager, CheckpointKind checkpointKind, string[] edgeTags, long payloadSizeBytes, long timeInQueueMs)
+        /// <param name="parent">TODO</param>
+        internal void SetCheckpoint(DataStreamsManager manager, CheckpointKind checkpointKind, string[] edgeTags, long payloadSizeBytes, long timeInQueueMs, PathwayContext? parent)
         {
-            PathwayContext = manager.SetCheckpoint(PathwayContext, checkpointKind, edgeTags, payloadSizeBytes, timeInQueueMs);
-        }
-
-        /// <summary>
-        /// Merges two DataStreams <see cref="PathwayContext"/>
-        /// Should be called when a pathway context is extracted from an incoming span
-        /// Used to merge contexts in a "fan in" scenario.
-        /// </summary>
-        internal void MergePathwayContext(PathwayContext? pathwayContext)
-        {
-            if (pathwayContext is null)
-            {
-                return;
-            }
-
-            if (PathwayContext is null)
-            {
-                PathwayContext = pathwayContext;
-                return;
-            }
-
-            // This is purposely not thread safe
-            // The code randomly chooses between the two PathwayContexts.
-            // If there is a race, then that's okay
-            // Randomly select between keeping the current context (0) or replacing (1)
-            if (ThreadSafeRandom.Shared.Next(2) == 1)
```

Comment:
this random was introduced [here](https://github.com/DataDog/dd-trace-dotnet/pull/3174/files#r980446221), with the justification that it was done like this in Java, but I couldn't find this code in the java tracer.
Also, the justification is that we'd want to select at random on fan-in to have all the sources _on average_. But on fan-in, we still want to have all the consume checkpoints matching the produce pathways, there is no problem here.
Where we'd need a random could be afterwards, on a subsequent produce. But I think just the natural randomness from the order of messages arriving should be enough to ensure we get all sources.

Reply:
FWIW: here is the code in the java tracer: https://github.com/DataDog/dd-trace-java/blob/0dc091aa3055de083eff6701336161ecaf710133/dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java#L647-L663

Reply:
Ah thanks, interesting... from what I see, this code is never called by our instrumentations. The only codepath leading to it is `setConsumeCheckpoint` -> `mergePathwayContextIntoSpan` -> `mergePathwayContext`
and `setConsumeCheckpoint` has no usage within our codebase (even in tests). I think it's meant for users checkpointing manually.

And even here, I don't think it's what we want. It has the same problem as dotnet, where if we consume twice in the same span, the second time, we're going to randomly choose between the previous pathway _after the consume_ and the pathway we read from the incoming message.
What we'd want, is for a potential subsequent produce to take one of the two post-consume-checkpoint pathways randomly. The random is not in the right place.